src/main/java/graphql/execution/ResolveType.java

```diff
@@ -19,9 +20,9 @@
 @Internal
 public class ResolveType {
 
-
     public GraphQLObjectType resolveType(ExecutionContext executionContext, MergedField field, Object source, ExecutionStepInfo executionStepInfo, GraphQLType fieldType, Object localContext) {
-        GraphQLObjectType resolvedType;
+        Assert.assertTrue(fieldType instanceof GraphQLInterfaceType || fieldType instanceof GraphQLUnionType,
+                () -> "The passed in fieldType MUST be an interface or union type : " + fieldType.getClass().getName());
```

Comment:
Now only interfaces and unions into this method.  The outer code does the object check.

I know its more self contained if perhaps did all three but performance is performance.  Not only do we avoid a env object allocation - we avoid a method call as well.